### PR TITLE
bound on recursion

### PR DESCRIPTION
**Pull request by weichweich**
## Checklist:

- [ ] I have verified that the code works
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: pallets/delegation/src/lib.rs [hunk 1]
```diff
@@ -329,35 +332,57 @@ impl<T: Trait> Module<T> {
...
 		// check if already revoked
 		if !delegation_node.revoked {
-			// set revoked flag and store delegation node
-			delegation_node.revoked = true;
-			<Delegations<T>>::insert(*delegation, delegation_node);
-			// deposit event that the delegation has been revoked
-			Self::deposit_event(RawEvent::DelegationRevoked(sender.clone(), *delegation));
-			// revoke all children recursively
-			Self::revoke_children(delegation, sender)?;
+			// fist revoke all children recursively
```
Comment (wischli) [inline pallets/delegation/src/lib.rs:347]: Minor typo here
```suggestion
			// first revoke all children recursively
```

File: pallets/delegation/src/lib.rs [hunk 2]
```diff
@@ -329,35 +332,57 @@ impl<T: Trait> Module<T> {
...
+			revocations += Self::revoke_children(delegation, sender, max_revocations)?;
+
+			// if we run out of revocation gas, we only revoke children. The tree will be changed but still valid.
```
Comment (wischli) [inline pallets/delegation/src/lib.rs:350]: Missing a word here
```suggestion
			// if we run out of revocation gas, we only revoke children. The tree will be changed but is still valid.
```

File: pallets/delegation/src/lib.rs [hunk 3]
```diff
@@ -130,6 +130,7 @@ decl_error! {
 		ParentNotFound,
 		UnauthorizedRevocation,
 		UnauthorizedDelegation,
+		ExeededBounds,
```
Comment (wischli) [inline pallets/delegation/src/lib.rs:133]: Typo here. I'd also suggest to get a more descriptive name like
```suggestion
		ExceededRevocationBounds,
```

**Review by wischli — CHANGES_REQUESTED** [review]
Three typos and I think we should also add a unit test for `ExceedRevocationBounds`.

File: pallets/delegation/src/lib.rs (hunk 3 above)
Comment (weichweich) [reply pallets/delegation/src/lib.rs]: fixed in 4966e4b

**Comment by weichweich** [comment]
> Three typos and I think we should also add a unit test for `ExceedRevocationBounds`.

good catch! Totally forgot that! There was even a bug there! The root was always revoked even if there where no revocations remaining, leading to an invalid delegation tree! 😨 Fixed in 4966e4b...

File: pallets/delegation/src/tests.rs [hunk 4]
```diff
@@ -442,17 +445,26 @@ fn check_add_and_revoke_delegations() {
...
 			),
 			Error::<Test>::RootNotFound
 		);
 		assert_noop!(
-			Delegation::revoke_root(Origin::signed(account_hash_bob), id_level_0),
-			Error::<Test>::UnauthorizedRevocation
+			Delegation::revoke_root(Origin::signed(account_hash_bob), id_level_0, 1),
+			Error::<Test>::UnauthorizedRevocation,
+		);
+		let res =
```
Comment (wischli) [inline pallets/delegation/src/tests.rs:457]: Why not the usual check?

```suggestion
		assert_noop!(Delegation::revoke_root(Origin::signed(account_hash_alice.clone()), id_level_0, 0), Error::<Test>::ExceededRevocationBounds);
```

Comment (weichweich) [reply pallets/delegation/src/tests.rs]: forgot that this exists. f8f7e8c

**Review by wischli — APPROVED** [review]
LGTM!

**Pull request merged**